src/Workspaces/Remote/ServiceHub/Host/RemoteSolutionCache.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+/// <summary>
+/// LRU cache of checksum+solution pairs.  Used to keep track of the last few solutions the remote server knows about,
+/// helping to avoid unnecessary syncs/recreations of those solutions while many requests are coming into the server.
+/// Not threadsafe.  Only use while under a lock.
+/// </summary>
+internal sealed class RemoteSolutionCache<TChecksum, TSolution>
+    where TChecksum : struct
+    where TSolution : class
+{
+    /// <summary>
+    /// The max number of solution instances we'll hold onto at a time.
+    /// </summary>
+    private readonly int _maxCapacity;
+
+    /// <summary>
+    /// The total history kept.  Used to record telemetry about how useful it would be to increase the max capacity.
+    /// </summary>
+    private readonly int _totalHistory;
+
+    /// <summary>
+    /// Keep track of when we find a checksum in the history, but we've dropped the solution for it.  This will help
+    /// us determine what benefit we would get from expanding this cache.
+    /// </summary>
+    private readonly HistogramLogAggregator<int>.HistogramCounter _cacheMissCounter;
+
+    /// <summary>
+    /// The number of times we successfully found a solution.
+    /// </summary>
+    private int _cacheHits;
+
+    /// <summary>
+    /// The number of times we failed to find a solution, but could have found it if we cached more items (up to
+    /// TotalHistory).  When this happens, we also store in <see cref="_cacheMissCounter"/> which bucket it was
+    /// found in to help us decide what a good cache value is.
+    /// </summary>
+    private int _cacheMissesInHistory;
+
+    /// <summary>
+    /// The number of times we failed to find a solution, and would not have found it even if we didn't cache more items.
+    /// </summary>
+    private int _cacheMissesNotInHistory;
+
+    /// <summary>
+    /// The list of checksum+solution pairs.  Note: only the first <see cref="_maxCapacity"/> items will actually point
+    /// at a non-null solution.  The ones after that will point at <see langword="null"/>.  We store both so that we can
+    /// collect useful telemetry on how much benefit we would get by having a larger history.
+    /// </summary>
+    private readonly LinkedList<CacheNode> _cacheNodes = new();
```

Comment:
NOte: this is the only real data in the type.  Everything above this is telemetry data.